core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -73,221 +112,807 @@ public static RecordState forId(byte id) {
+            return offsetState;
+        }
+
+        private InFlightState tryUpdateBatchState(RecordState newState, boolean incrementDeliveryCount) {
+            if (inFlightState == null) {
+                throw new IllegalStateException("The batch state update is not available as the offset state is maintained");
+            }
+            return inFlightState.tryUpdateState(newState, incrementDeliveryCount);
+        }
+
+        private InFlightState startBatchStateTransition(RecordState newState, boolean incrementDeliveryCount) {
+            if (inFlightState == null) {
+                throw new IllegalStateException("The batch state update is not available as the offset state is maintained");
+            }
+            return inFlightState.startStateTransition(newState, incrementDeliveryCount);
+        }
+
+        private void addGapOffsets(Long gapOffset) {
+            if (this.gapOffsets == null) {
+                this.gapOffsets = new HashSet<>();
+            }
+            this.gapOffsets.add(gapOffset);
+        }
+
+        private void addGapOffsets(List<Long> gapOffsets) {
+            if (this.gapOffsets == null) {
+                this.gapOffsets = new HashSet<>(gapOffsets);
+            } else {
+                this.gapOffsets.addAll(gapOffsets);
+            }
+        }
+
+        private void addOffsetState(long baseOffset, long lastOffset, RecordState state, boolean incrementDeliveryCount) {
+            maybeInitializeOffsetStateUpdate();
+            // The offset state map is already initialized hence update the state of the offsets.
+            for (long offset = baseOffset; offset <= lastOffset; offset++) {
+                offsetState.get(offset).tryUpdateState(state, incrementDeliveryCount);
+            }
+        }
+
+        private void addOffsetState(Map<Long, RecordState> stateMap, boolean incrementDeliveryCount) {
+            maybeInitializeOffsetStateUpdate();
+            // The offset state map is already initialized hence update the state of the offsets.
+            stateMap.forEach((offset, state) -> offsetState.get(offset).tryUpdateState(state, incrementDeliveryCount));
+        }
+
+        private void maybeInitializeOffsetStateUpdate() {
+            if (offsetState == null) {
+                offsetState = new ConcurrentSkipListMap<>();
+                // The offset state map is not initialized hence initialize the state of the offsets
+                // from the base offset to the last offset. Mark the batch inflightState to null as
+                // the state of the records is maintained in the offset state map now.
+                for (long offset = this.baseOffset; offset <= this.lastOffset; offset++) {
+                    if (gapOffsets != null && gapOffsets.contains(offset)) {
+                        // Directly move the record to archived if gap offset is already known.
+                        offsetState.put(offset, new InFlightState(RecordState.ARCHIVED, 0));
```

Comment:
I was expecting that you'd just omit this offset from the map. Either way works I suppose.

Reply:
Yeah removing it also make sense, I ll keep it as it is for now but may be better to remove it as well.